src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java

```diff
@@ -739,6 +747,326 @@ public boolean apply(@Nullable FlowNode input) {
+                actions.add(a);
+            }
+        }
+        return actions;
+    }
+
+    /**
+     * Needs at least Jenkins version 2.145 which supports {@link ExecutorStepExecution.PlaceholderTask#getAffinityKey()}
+     */
+    @Issue("JENKINS-36547")
+    @Test public void reuseNodeFromPreviousRun() {
+        story.then(r -> {
+            Assume.assumeFalse("Needs Jenkins version >= 2.145", Jenkins.getVersion().isOlderThan(new VersionNumber("2.145")));
+            for (int i = 0; i < 5; ++i) {
+                DumbSlave slave = r.createOnlineSlave();
+                slave.setLabelString("foo bar");
+            }
+
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "demo");
+            p.setDefinition(new CpsFlowDefinition("node('foo') {\n" +
+                    "}\n", true));
+
+            WorkflowRun run = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
+            List<WorkspaceAction> workspaceActions = getWorkspaceActions(run);
+            assertEquals(workspaceActions.size(), 1);
+
+            String firstNode = workspaceActions.get(0).getNode();
+            assertNotEquals(firstNode, "");
+
+            WorkflowRun run2 = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
+            workspaceActions = getWorkspaceActions(run2);
+            assertEquals(workspaceActions.size(), 1);
+            assertEquals(workspaceActions.get(0).getNode(), firstNode);
+        });
+    }
+
+    private Map<String, StringWriter> mapNodeNameToLogText(WorkflowRun workflowRun) throws java.io.IOException{
+        FlowGraphWalker walker = new FlowGraphWalker(workflowRun.getExecution());
+        Map<String, StringWriter> workspaceActionToLogText = new HashMap<>();
+        for (FlowNode n : walker) {
+            if (n instanceof StepAtomNode) {
+                StepAtomNode atomNode = (StepAtomNode) n;
+                if (atomNode.getDescriptor() instanceof EchoStep.DescriptorImpl) {
+                    // we're searching for the echo only...
+                    LogAction l = atomNode.getAction(LogAction.class);
+                    if (l != null) {
+                        // Only store the log if there was no workspace action involved... (e.g. from echo)
+                        List<? extends BlockStartNode> enclosingBlocks = atomNode.getEnclosingBlocks();
+                        for (BlockStartNode startNode : enclosingBlocks) {
+                            WorkspaceAction a = startNode.getAction(WorkspaceAction.class);
+                            if (a != null) {
+                                String nodeName = a.getNode();
+                                if (!workspaceActionToLogText.containsKey(nodeName)) {
+                                    workspaceActionToLogText.put(nodeName, new StringWriter());
+                                }
+                                StringWriter writer = workspaceActionToLogText.get(nodeName);
+                                l.getLogText().writeLogTo(0, writer);
```

Comment:
Wait, why are we writing to the build log from inside a test?

Reply:
@jglick: Actually what happens here is reading the build log and storing it inside `writer`. Or was it that what you were asking for?
This helper just provides the build log to the tests so it can be checked for correctness. But maybe there's a better way of doing that? So far I could not come up with a better idea.

Reply:
`JenkinsRule.assertLogContains`?

Reply:
Thanks. Actually I forgot to mention that I need to check for the console log which belongs to a specific node block to see whether it has been executed on the expected node. First think I tried was to somehow retrieve the node name and check that.

After a while I thought it'd be better to have a small helper which will split the console log and map the node names to their corresponding log text.
Now that I think about it I probably could have also just retrieved the node name using `getContext()` and `echo` it to the console where required.

However since the solution in place will be little bit more flexible I'd prefer that one.
I added some doxygen comment to hopefully clarify about that method's purpose.

Ok to keep it?